visa/views.py

```diff
@@ -0,0 +1,80 @@
+from django.conf import settings
+from django.contrib.auth.mixins import LoginRequiredMixin
+from django.core.mail.message import EmailMessage
+from django.shortcuts import redirect
+from django.utils import timezone
+from django.views.generic import CreateView
+
+from visa.forms.visa_letter import VisaLetterForm
+
+from .models import VisaInvitationLetter
+
+
+class VisaLetterCreateView(LoginRequiredMixin, CreateView):
+    model = VisaInvitationLetter
+    form_class = VisaLetterForm
+    template_name = "visa/visa_letter_form.html"
+
+    def get_success_url(self):
+        return self.request.path
+
+    def form_valid(self, form):
+        form.instance.user = self.request.user
+        form.instance.status = "pending"
+        response = super().form_valid(form)
+
+        self.request.session["visa_form_success"] = True
+
+        return response
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        context["current_datetime"] = timezone.now().strftime("%Y-%m-%d %H:%M:%S")
+        context["current_user"] = self.request.user.username
+
+        context["form_submitted"] = self.request.session.pop("visa_form_success", False)
+
+        context["error_message"] = self.request.session.pop("visa_error", None)
+        context["success_message"] = self.request.session.pop("visa_success", None)
+        context["info_message"] = self.request.session.pop("visa_info", None)
+
+        try:
+            existing_letter = self.request.user.visa_letter
+            context["existing_letter"] = existing_letter
+            context["has_existing_letter"] = True
+            context["latest_letter"] = existing_letter
+        except VisaInvitationLetter.DoesNotExist:
+            context["has_existing_letter"] = False
+
+        return context
+
+    def post(self, request, *args, **kwargs):
+        if "resend_email" in request.POST:
+            try:
+                letter = request.user.visa_letter
+            except VisaInvitationLetter.DoesNotExist:
+                request.session["visa_error"] = (
```

Comment:
why not just use a "get object or 404". Or  `messages` ? 

All the session wrangling seems a bit hacky. It can work, but it's like using a screwdriver as a hammer.